### PR TITLE
Issue #28: Generic HTML support based on filter settings

### LINKED ISSUE #28: Integrate Generic HTML Support for Backwards-Compatibility
The [Generic HTML Support](https://ckeditor.com/docs/ckeditor5/latest/features/html/general-html-support.html) plugin provides support for tags that are otherwise unsupported by CKEditor. For example CKEditor doesn't know how to handle a generic `<div>`, `<span>`, or `<cite>` elements, as well as any arbitrary attributes like `style` or `id`.

For backwards-compatibility with existing content, CKEditor 5 module should allow a pretty wide-ranging set of attributes and tags, based on the filter configuration (if the tag filter is enabled).

### PR DESCRIPTION
**Pull request by indigoxela**
Fixes #28 – possibly.

### CODE REVIEW

File: ckeditor5.module [hunk 1]
```diff
@@ -958,10 +966,15 @@ function ckeditor5_get_acf_settings($format) {
...
+      //
+      // @todo Placeholder "*" is not working. The filter still gets applied to
+      // markup after saving. But for CKEditor5 it's meaningless.
+      // @see ckeditor5.js
       $disallowed_attributes = array_filter($attributes, function($value) {
         return $value === FALSE;
       });
       if (count($disallowed_attributes)) {
+        $disallowed[$tag]['name'] = $tag;
         // No need to blacklist the 'class' or 'style' attributes; CKEditor
```
Comment (olafgrabienski) [inline ckeditor5.module:978]: Maybe change "blacklist" to "disallow", as working on this file?

Comment (indigoxela) [reply ckeditor5.module:978]: When reading the mail, I got confused (I never use the term "blacklist" nowadays).

But now I get it – it's not my code. :laughing: Sure, can update that.

File: ckeditor5.module [hunk 2]
```diff
@@ -871,18 +878,18 @@ function ckeditor5_get_config($format, array $existing_settings) {
...
 function ckeditor5_get_acf_settings($format) {
   $html_restrictions = filter_format_allowed_html($format);
 
-  // When all HTML is allowed, also set allowedContent to true and
-  // disallowedContent to false.
+  // When all HTML is allowed, we return empty arrays. This case gets handled in
+  // js/ckeditor5.js.
   if ($html_restrictions === TRUE) {
-    return array(TRUE, FALSE);
+    return array(array(), array());
```
Comment (quicksketch) [inline ckeditor5.module:892]: I like this change.

File: ckeditor5.module [hunk 3]
```diff
@@ -1014,7 +1022,13 @@ function ckeditor5_get_acf_settings($format) {
         return $value === FALSE;
       });
       if (count($disallowed_attributes)) {
-        // No need to blacklist the 'class' or 'style' attributes; CKEditor
+        // This pattern has no meaning for CKEditor5.
+        // @see js/ckeditor5.js
```
Comment (quicksketch) [inline ckeditor5.module]: Can you elaborate on this condition? Does this just mean that you can't have `*` as a disallowed attribute?

File: js/ckeditor5.js [hunk 4]
```diff
@@ -28,6 +28,25 @@
 
       editorSettings.licenseKey = '';
 
+      // If filter_html is turned on, we prevent on* attributes.
+      if (editorSettings.htmlSupport.allow.length) {
```
Comment (quicksketch) [inline js/ckeditor5.js]: Could we wrap both these checks to verify that HTML Support is enabled before adding any of these settings? In the event that `htmlSupport.GeneralHtmlSupport` is disabled through `hook_ckeditor5_config_alter()`.

File: ckeditor5.module [hunk 5]
```diff
@@ -871,18 +878,18 @@ function ckeditor5_get_config($format, array $existing_settings) {
...
- *   - the first value is the "allowedContent" setting: a well-formatted array
- *     or TRUE. The latter indicates that anything is allowed.
- *   - the second value is the "disallowedContent" setting: a well-formatted
- *     array or FALSE. The latter indicates that nothing is disallowed.
+ *   - Configuration array of allowed content rules used by the editor's
+ *     General HTML Support "allow" setting.
+ *   - Configuration array of disallowed content rules used by the editor's
+ *     General HTML Support "disallow" setting.
  */
 function ckeditor5_get_acf_settings($format) {
```
Comment (quicksketch) [inline ckeditor5.module]: Can we rename this function to match its new purpose? I'm surprised so much of this was reusable, it was just copied over from CKEditor 4. Maybe we name this function `ckeditor5_get_generic_html_settings()`?

Comment (indigoxela) [reply ckeditor5.module]: Good idea! :+1:

File: js/ckeditor5.js (hunk 4 above)
Comment (indigoxela) [reply js/ckeditor5.js]: Right, there's also ...config_alter(). Didn't think of that. :+1:

File: ckeditor5.module (hunk 3 above)
Comment (indigoxela) [reply ckeditor5.module]: We can't use "*" anywhere, as the MatcherPattern's supposed to be either a string (then it's literal matching), or a Regexp. I tried to update the comment to explain a bit better.

OK, we can use "*", of course, but it won't match anything. :wink:

**Pull request merged**